plugin tries to execute 'null/pdflatex'

#### Type of JetBrains IDE (IntelliJ, PyCharm, etc.) and version
IntelliJ IDEA Ultimate

#### Operating System 
Ubuntu 20.10


#### TeXiFy IDEA version
0.7.4  
I also tested this in 0.7.3. The error did not occur there.

#### What I did (steps to reproduce)
I added the suggested sdk at `/usr/bin` and create a template project. Then I created a run configuration by clicking the gutter icon left of `\begin{document}`.

#### Minimal example to reproduce the problem
```latex
\documentclass{article}
\begin{document}
    
\end{document}
```

#### Expected behavior
When running with `Ctrl-Shift-F10`, I expected the plugin to run the programm at `/usr/bin/pdflatex`.

#### Actual behavior
Following error ist shown:
```
Error running 'main':Cannot run program "null/pdflatex" (in directory "/home/user/project/src"): error=2, No such file or directory
```

The full stacktrace from the logs is:
```
Cannot run program "null/pdflatex" (in directory "/home/user/project/src"): error=2, No such file or directory
	at com.intellij.execution.configurations.GeneralCommandLine.createProcess(GeneralCommandLine.java:380)
	at com.intellij.execution.process.OSProcessHandler.startProcess(OSProcessHandler.java:101)
	at com.intellij.execution.process.OSProcessHandler.<init>(OSProcessHandler.java:49)
	at com.intellij.execution.process.KillableProcessHandler.<init>(KillableProcessHandler.java:37)
	at nl.hannahsten.texifyidea.run.latex.LatexCommandLineState.createHandler(LatexCommandLineState.kt:77)
	at nl.hannahsten.texifyidea.run.latex.LatexCommandLineState.startProcess(LatexCommandLineState.kt:57)
	at com.intellij.execution.configurations.CommandLineState.execute(CommandLineState.java:61)
	at com.intellij.execution.runners.DefaultRunProgramRunner$execute$1.invoke(DefaultRunProgramRunner.kt:29)
	at com.intellij.execution.runners.DefaultRunProgramRunner$execute$1.invoke(DefaultRunProgramRunner.kt:12)
	at com.intellij.execution.impl.ExecutionManagerImpl$startRunProfile$1.invoke(ExecutionManagerImpl.kt:155)
	at com.intellij.execution.impl.ExecutionManagerImpl$startRunProfile$1.invoke(ExecutionManagerImpl.kt:65)
	at com.intellij.execution.impl.ExecutionManagerImpl$doStartRunProfile$startRunnable$1.run(ExecutionManagerImpl.kt:216)
	at com.intellij.openapi.application.TransactionGuardImpl.runWithWritingAllowed(TransactionGuardImpl.java:216)
	at com.intellij.openapi.application.TransactionGuardImpl.access$200(TransactionGuardImpl.java:24)
	at com.intellij.openapi.application.TransactionGuardImpl$2.run(TransactionGuardImpl.java:199)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:822)
	at com.intellij.openapi.application.impl.ApplicationImpl.lambda$invokeLater$4(ApplicationImpl.java:324)
	at com.intellij.openapi.application.impl.FlushQueue.doRun(FlushQueue.java:85)
	at com.intellij.openapi.application.impl.FlushQueue.runNextEvent(FlushQueue.java:134)
	at com.intellij.openapi.application.impl.FlushQueue.flushNow(FlushQueue.java:47)
	at com.intellij.openapi.application.impl.FlushQueue$FlushNow.run(FlushQueue.java:190)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:313)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:776)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:727)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:746)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:976)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:843)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$8(IdeEventQueue.java:454)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:773)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$9(IdeEventQueue.java:453)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:822)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:501)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
Caused by: java.io.IOException: Cannot run program "null/pdflatex" (in directory "/home/user/project/src"): error=2, No such file or directory
	at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1128)
	at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1071)
	at com.intellij.execution.configurations.GeneralCommandLine.startProcess(GeneralCommandLine.java:453)
	at com.intellij.execution.configurations.GeneralCommandLine.createProcess(GeneralCommandLine.java:376)
	... 40 more
Caused by: java.io.IOException: error=2, No such file or directory
	at java.base/java.lang.ProcessImpl.forkAndExec(Native Method)
	at java.base/java.lang.ProcessImpl.<init>(ProcessImpl.java:340)
	at java.base/java.lang.ProcessImpl.start(ProcessImpl.java:271)
	at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1107)
	... 43 more
```

--- a/src/nl/hannahsten/texifyidea/completion/LatexCommandProvider.kt
+++ b/src/nl/hannahsten/texifyidea/completion/LatexCommandProvider.kt
@@ -39,6 +39,7 @@ class LatexCommandProvider internal constructor(private val mode: LatexMode) :
     CompletionProvider<CompletionParameters>() {
 
     companion object {
+
         /** Cache for commands which are indexed and which should be added to the autocompletion. */
         val indexedCommands = mutableSetOf<LookupElementBuilder>()
     }

--- a/src/nl/hannahsten/texifyidea/completion/LatexCommandProvider.kt
+++ b/src/nl/hannahsten/texifyidea/completion/LatexCommandProvider.kt
@@ -38,6 +38,11 @@ import java.util.stream.Collectors
 class LatexCommandProvider internal constructor(private val mode: LatexMode) :
     CompletionProvider<CompletionParameters>() {
 
+    companion object {
+        /** Cache for commands which are indexed and which should be added to the autocompletion. */
+        val indexedCommands = mutableSetOf<LookupElementBuilder>()
+    }
+
     override fun addCompletions(
         parameters: CompletionParameters,
         context: ProcessingContext,
@@ -77,7 +82,16 @@ class LatexCommandProvider internal constructor(private val mode: LatexMode) :
         }
     }
 
+    /**
+     * Add all indexed commands to the autocompletion.
+     */
     private fun addIndexedCommands(result: CompletionResultSet, parameters: CompletionParameters) {
+        // Use cache if available
+        if (indexedCommands.isNotEmpty()) {
+            result.addAllElements(indexedCommands)
+            return
+        }
+
         val commands = mutableSetOf<LookupElementBuilder>()
         FileBasedIndex.getInstance().getAllKeys(LatexExternalCommandIndex.id, parameters.editor.project ?: return)
             .forEach { cmdWithSlash ->
@@ -98,6 +112,7 @@ class LatexCommandProvider internal constructor(private val mode: LatexMode) :
                         .forEach { commands.add(it) }
                 }
             }
+        indexedCommands.addAll(commands)
         result.addAllElements(commands)
     }
 

--- a/src/nl/hannahsten/texifyidea/reference/InputFileReference.kt
+++ b/src/nl/hannahsten/texifyidea/reference/InputFileReference.kt
@@ -67,7 +67,7 @@ class InputFileReference(
         // Find the sources root of the current file.
         // findRootFile will also call getImportPaths, so that will be executed twice
         val rootFile = givenRootFile ?: element.containingFile.findRootFile().virtualFile
-        val rootDirectory = rootFile.parent ?: return null
+        val rootDirectory = rootFile?.parent ?: return null
 
         var targetFile: VirtualFile? = null
 

--- a/src/nl/hannahsten/texifyidea/settings/sdk/MiktexWindowsSdk.kt
+++ b/src/nl/hannahsten/texifyidea/settings/sdk/MiktexWindowsSdk.kt
@@ -33,7 +33,7 @@ class MiktexWindowsSdk : LatexSdk("MiKTeX Windows SDK") {
     override fun suggestHomePaths(): MutableCollection<String> {
         val results = mutableSetOf<String>()
         val paths = "where pdflatex".runCommand()
-        if (paths != null) {
+        if (paths != null && !paths.contains("Could not find")) { // Full output is INFO: Could not find files for the given pattern(s).
             paths.split("\r\n").forEach { path ->
                 val index = path.findLastAnyOf(setOf("miktex\\bin"))?.first ?: path.length - 1
                 results.add(path.substring(0, index))

--- a/src/nl/hannahsten/texifyidea/settings/sdk/NativeTexliveSdk.kt
+++ b/src/nl/hannahsten/texifyidea/settings/sdk/NativeTexliveSdk.kt
@@ -58,7 +58,11 @@ class NativeTexliveSdk : TexliveSdk("Native TeX Live SDK") {
         return """TeX Live (\d\d\d\d\/.+)""".toRegex().find(LatexSdkUtil.pdflatexVersionText)?.value ?: "Unknown version"
     }
 
-    override fun getDefaultDocumentationUrl(sdk: Sdk): String? {
+    override fun getDefaultDocumentationUrl(sdk: Sdk): String {
         return "$texmfDistPath/doc"
+    }
+
+    override fun getExecutableName(executable: String, homePath: String): String {
+        return "$homePath/$executable"
     }
 }

--- a/src/nl/hannahsten/texifyidea/settings/sdk/TexliveSdk.kt
+++ b/src/nl/hannahsten/texifyidea/settings/sdk/TexliveSdk.kt
@@ -93,6 +93,6 @@ open class TexliveSdk(name: String = "TeX Live SDK") : LatexSdk(name) {
     override fun getExecutableName(executable: String, homePath: String): String {
         // Get base path of LaTeX distribution
         val basePath = LatexSdkUtil.getPdflatexParentPath(homePath)
-        return "$basePath/$executable"
+        return if (basePath != null) "$basePath/$executable" else executable
     }
 }

